### PR TITLE
NativeToken Minting Functionalities

### PR DESCRIPTION
**Pull request by dushyantb**

### CODE REVIEW

File: src/Lachain.Core/Blockchain/SystemContracts/Interface/Lrc20Interface.cs [hunk 1]
```diff
@@ -13,6 +13,9 @@ public class Lrc20Interface : IContractInterface
         public const string MethodName = "name()";
         public const string MethodDecimals = "decimals()";
         public const string MethodSymbol = "symbol()";
+        public const string MethodMint = "mint(address,uint256)";
+        public const string MethodSetAllowedSupply = "setAllowedSupply(uint256)";
+        public const string MethodGetAllowedSupply = "getAllowedSupply()";
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/Interface/Lrc20Interface.cs:18]: I think we should return uint256 here

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs [hunk 2]
```diff
@@ -128,6 +147,48 @@ public ExecutionStatus Approve(UInt160 spender, UInt256 amount, SystemContractEx
...
         }
+        
+        [ContractMethod(Lrc20Interface.MethodSetAllowedSupply)]
+        public ExecutionStatus SetAllowedSupply(Money amount, SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            if(!frame.InvocationContext.Sender.Equals(_mintCntrlAdd))
+                return ExecutionStatus.ExecutionHalted;
+            
+            if(_allowedSupply + amount > _maxSupply)
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:158]: We should check just amount:
```suggestion
            if(amount > _maxSupply)
```

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs [hunk 3]
```diff
@@ -128,6 +147,48 @@ public ExecutionStatus Approve(UInt160 spender, UInt256 amount, SystemContractEx
...
+                return ExecutionStatus.ExecutionHalted;
+            
+            _allowedSupply = amount;
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:161]: Also we should check that new supply is bigger than current total supply

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs [hunk 4]
```diff
@@ -128,6 +147,48 @@ public ExecutionStatus Approve(UInt160 spender, UInt256 amount, SystemContractEx
...
+        public ExecutionStatus Mint(UInt160 address, Money amount, SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            if(!frame.InvocationContext.Sender.Equals(_minterAdd))
+                return ExecutionStatus.ExecutionHalted;
+            
+            var currentBalance =  _context.Snapshot?.Balances.GetBalance(address);
+            if (currentBalance is null) return ExecutionStatus.ExecutionHalted;
+            
+            if(currentBalance + amount > _maxSupply || currentBalance + amount > _allowedSupply)
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:184]: We should check totalSupply here,  not current address balance

**Review by sgladkov — CHANGES_REQUESTED** [review]

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs (hunk 4 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:184]: incorporated

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs (hunk 3 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:161]: incorporated

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs (hunk 2 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:158]: incorporated

File: src/Lachain.Core/Blockchain/SystemContracts/Interface/Lrc20Interface.cs (hunk 1 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/Interface/Lrc20Interface.cs:18]: we are not returning any values here, we are setting the actual values in the `frame.ReturnValue` and returning `ExecutionStatus` which I string

**Review by dushyantb — COMMENTED** [review]
pushed the new changes

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs [hunk 5]
```diff
@@ -22,10 +24,27 @@ public class NativeTokenContract : ISystemContract
...
+        private static byte[] _minterPubKey = CryptoUtils.EncodeCompressed(_minterKeyPair.PublicKey);
+        private static UInt160 _minterAdd = Crypto.ComputeAddress(_minterPubKey).ToUInt160();
+
+        private static EcdsaKeyPair _mintCntrlKeyPair =
+            new EcdsaKeyPair("0xE83385AF76B2B1997326B567461FB73DD9C27EAB9E1E86D26779F4650C5F2B75".HexToBytes()
+                .ToPrivateKey());
+        private static byte[] _mintCntrlPubKey = CryptoUtils.EncodeCompressed(_mintCntrlKeyPair.PublicKey);
+        private static UInt160 _mintCntrlAdd = Crypto.ComputeAddress(_mintCntrlPubKey).ToUInt160();
+
+        private Money _allowedSupply = Money.Zero;
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:45]: This values should be stored in storage,  in other case we will have a zero allowed supply every time we recreate the contract instance

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs [hunk 6]
```diff
@@ -128,6 +147,50 @@ public ExecutionStatus Approve(UInt160 spender, UInt256 amount, SystemContractEx
...
+        public ExecutionStatus SetAllowedSupply(Money amount, SystemContractExecutionFrame frame)
+        {
+            frame.UseGas(GasMetering.NativeTokenApproveCost);
+            if(!frame.InvocationContext.Sender.Equals(_mintCntrlAdd))
+                return ExecutionStatus.ExecutionHalted;
+            
+            if(amount > _maxSupply)
+                return ExecutionStatus.ExecutionHalted;
+            
+            if(amount <= _allowedSupply)
```
Comment (sgladkov) [inline src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs:161]: No,  there is some misunderstanding here,  it is possible to decrease allowed supply,  but we can't make it less than current **total** supply

**Review by sgladkov — CHANGES_REQUESTED** [review]

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs (hunk 5 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs]: added respective methods in `BalanceSnapshot.cs`

File: src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs (hunk 6 above)
Comment (dushyantb) [reply src/Lachain.Core/Blockchain/SystemContracts/NativeTokenContract.cs]: incorporated

**Review by sgladkov — APPROVED** [review]

**Pull request merged**